components/vc/front/src/torque.c

```diff
@@ -47,13 +54,167 @@ static struct
+    torque_data.torque_control_request_active = (CANRX_get_signal(VEH, SWS_requestLaunchControl, &launch_control_requested) != CANRX_MESSAGE_SNA) &&
+                                                (launch_control_requested == CAN_DIGITALSTATUS_ON);
+    const bool launch_control_request_rising = !launch_control_request_was_active && torque_data.torque_control_request_active;
+
+    switch (torque_data.launch_control_state)
+    {
+        case LAUNCHCONTROL_REJECTED:
+            if (drv_timer_getState(&torque_data.launch_control_timer) == DRV_TIMER_EXPIRED)
+            {
+                torque_data.launch_control_state = LAUNCHCONTROL_INACTIVE;
+            }
+            __attribute__((fallthrough));
+        case LAUNCHCONTROL_INACTIVE:
+            if (launch_control_request_rising)
+            {
+                if ((brake_position > LAUNCH_CONTROL_BRAKE_THRESHOLD) &&
+                    (accelerator_position < LAUNCH_CONTROL_THROTTLE_START_CUTOFF))
+                {
+                    torque_data.launch_control_state = LAUNCHCONTROL_HOLDING;
+                    drv_timer_stop(&torque_data.launch_control_timer);
+                }
+                else
+                {
+                    torque_data.launch_control_state = LAUNCHCONTROL_REJECTED;
+                    drv_timer_start(&torque_data.launch_control_timer, LAUNCH_CONTROL_REJECTED_MS);
+                }
+            }
+            break;
+        case LAUNCHCONTROL_HOLDING:
+            if (brake_position < LAUNCH_CONTROL_BRAKE_THRESHOLD)
+            {
+                torque_data.launch_control_state = LAUNCHCONTROL_INACTIVE;
+            }
+            else if (accelerator_position > LAUNCH_CONTROL_THROTTLE_THRESHOLD)
+            {
+                drv_timer_start(&torque_data.launch_control_timer, LAUNCH_CONTROL_SETTLING_MS);
+                torque_data.launch_control_state = LAUNCHCONTROL_SETTLING;
+            }
+            break;
+        case LAUNCHCONTROL_SETTLING:
+            if (drv_timer_getState(&torque_data.launch_control_timer) == DRV_TIMER_EXPIRED)
+            {
+                torque_data.launch_control_state = LAUNCHCONTROL_PRELOAD;
+                drv_timer_stop(&torque_data.launch_control_timer);
+            }
+            else if (accelerator_position < LAUNCH_CONTROL_THROTTLE_THRESHOLD)
+            {
+                torque_data.launch_control_state = LAUNCHCONTROL_HOLDING;
+            }
+            else if (brake_position < LAUNCH_CONTROL_BRAKE_THRESHOLD)
+            {
+                torque_data.launch_control_state = LAUNCHCONTROL_INACTIVE;
+            }
+            break;
+        case LAUNCHCONTROL_PRELOAD:
+            if (accelerator_position < LAUNCH_CONTROL_THROTTLE_THRESHOLD)
+            {
+                torque_data.launch_control_state = LAUNCHCONTROL_INACTIVE;
+            }
+            else if (brake_position < LAUNCH_CONTROL_BRAKE_THRESHOLD)
```

Comment:
you should set a minimum threshold here too. Like make sure the driver releases the brakes below 5% or smth, not just to 29%. So do like a `LAUNCH_CONTROL_BRAKE_THRESHOLD_ENTER` and `LAUNCH_CONTROL_BRAKE_THRESHOLD_LAUNCH`

Reply:
Thats valid, so only enter launch once you have gotten under 5%?